rclrs/build.rs

```diff
@@ -27,10 +28,65 @@ fn main() {
     let ament_prefix_var_name = "AMENT_PREFIX_PATH";
     let ament_prefix_var = env::var(ament_prefix_var_name);
 
+    // #############
+    // # ALGORITHM #
+    // #############
+    //
+    // For each prefix in ${AMENT_PREFIX_PATH}:
+    // Search through ament index at ${prefix}/share/ament_index/resource_index/packages/ to find packages to include
+    // The include root will be located at either:
+    // - ${prefix}/include/ (old style)
+    // - ${prefix}/include/package_name (new style)
+    // - ${prefix}/include/CycloneDDS (special case, match for this)
+    // End of loop
+    // Compiled libraries are always at ${prefix}/lib
+    //
+    // See REP 122 for more details: https://www.ros.org/reps/rep-0122.html#filesystem-layout
+
     if let Ok(ament_prefix_paths) = ament_prefix_var {
-        for ament_prefix_path in ament_prefix_paths.split(':') {
-            builder = builder.clang_arg(format!("-I{}/include", ament_prefix_path));
-            println!("cargo:rustc-link-search=native={}/lib", ament_prefix_path);
+        for ament_prefix_path in ament_prefix_paths.split(':').map(PathBuf::from) {
+            // Locate the ament index
+            let ament_index = ament_prefix_path.join("share/ament_index/resource_index/packages");
+            if !ament_index.is_dir() {
+                continue;
+            }
+
+            // Base directory all headers will be
+            let include_dir = ament_prefix_path.join("include");
+
+            // Including the old-style packages
+            // todo(jhdcs): Remove this when old-style installation paths are fully removed
+            builder = builder.clang_arg(format!("-isystem{}", include_dir.to_string_lossy()));
+
+            // Search for and include new-style-converted package paths
+            for dir_entry in read_dir(&ament_index).unwrap().filter_map(|p| p.ok()) {
+                let package = dir_entry.file_name().to_string_lossy().to_string();
+                let project_include_dir = include_dir.join(&package);
+
+                if project_include_dir.is_dir() {
+                    // If the project_include_dir has another directory within it with the same name as the project
+                    // then this is a new-style installation path
+                    // todo(jhdcs): remove this check when no more packages use the old style
```

Comment:
Nit: I think this doesn't need to be a todo, since we still want to support Foxy and Galactic, right?